rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/History.scala

```diff
@@ -0,0 +1,375 @@
+
+          case head :+ Skip(affix, _) =>
+            Applicative[F]
+              .pure(
+                (
+                  currentPath.dropRight(affix.size.toInt),
+                  head,
+                  Skip(affix, Trie.hash(nextLastSeen)),
+                  currentAcc :+ nextLastSeen
+                )
+              )
+              .map(_.asLeft)
+          case Vector() => Applicative[F].pure(currentAcc :+ nextLastSeen).map(_.asRight)
+        }
+    }
+    FlatMap[F].tailRecM((path, rest, lastSeen, acc))(go)
+  }
+
+  private def rebalanceDelete(p1: KeyPath, r: Vector[Trie], a: Option[Trie]): F[List[Trie]] = {
+    type Params = (KeyPath, Vector[Trie], Option[Trie])
+    def go(params: Params): F[Either[Params, List[Trie]]] =
+      params match {
+        case (currentKey, init :+ Skip(affix, _), None) => // remove empty skip
+          Applicative[F].pure((currentKey.dropRight(affix.size.toInt), init, None)).map(_.asLeft)
+
+        case (currentKey, init :+ Skip(affix, _), Some(Skip(eaffix, ep))) => // merge skips
+          rehash(currentKey.dropRight(affix.size.toInt), init, Skip(affix ++ eaffix, ep), Nil)
+            .map(_.asRight)
+
+        case (keyInit :+ keyLast, init :+ Node(p), None) => // handle pointer block
+          pointerBlockStore.get(p).flatMap {
+            case Some(pointerBlock) if pointerBlock.countNonEmpty == 2 =>
+              // pointerBlock contains 1 more element, find it, wrap in skip and re-balance
+              def getOther =
+                pointerBlock
+                  .updated((toInt(keyLast), EmptyTrie) :: Nil)
+                  .toVector
+                  .zipWithIndex
+                  .filter(v => v._1 != EmptyTrie)
+                  .head
+
+              val (other, idx) = getOther
+              Applicative[F].pure((keyInit, init, Some(pointTo(toByte(idx), other)))).map(_.asLeft)
+            case Some(pb) => // pb contains 3+ elements, drop one, rehash
+              val updated = pb.updated((toInt(keyLast), EmptyTrie) :: Nil)
+              storePointerBlock(updated)
+                .flatMap(node => rehash(keyInit, init, node, Nil).map(_.asRight))
+            case None => Sync[F].raiseError(new RuntimeException("malformed trie"))
+          }
+
+        case (key, rest @ _ :+ (_: Node), Some(s: Skip)) =>
+          rehash(key, rest, s, Nil).map(_.asRight)
+
+        case (Nil, Vector(), Some(t)) => Applicative[F].pure(t :: Nil).map(_.asRight)
+
+        case (Nil, Vector(), None) => Applicative[F].pure(EmptyTrie :: Nil).map(_.asRight)
+      }
+    FlatMap[F].tailRecM((p1, r, a))(go)
+  }
+
+  def process(actions: List[HistoryAction[K]]): F[History[F, K]] =
+    // TODO this is an intermediate step to reproduce all the trie behavior
+    // will evolve to a fold based implementation with partial tries
+    actions
+      .foldLeftM(this.root) {
+        case (currentRoot, InsertAction(k, value)) =>
+          val currentPath = asBytes(k)
```

Comment:
I'd find `remainingPath` clearer - both here, and in the parameter name in traverseTeee